### ISSUE TITLE
 What do the elements in configs/model.json mean?

### ISSUE THREAD

**Issue by jungtaekyung1** [neosapience/mlp-singer#9]
I have successfully trained an example model using csd data. So, this time I am trying to learn with my own data.

However, when I try to train my own song data to the model, this error occurs.

```
/pytorch/aten/src/ATen/native/cuda/Indexing.cu:699: indexSelectLargeIndex: block: [379,0,0], thread: [123,0,0] Assertion `srcIndex < srcSelectDimSize` failed.
/pytorch/aten/src/ATen/native/cuda/Indexing.cu:699: indexSelectLargeIndex: block: [379,0,0], thread: [124,0,0] Assertion `srcIndex < srcSelectDimSize` failed.
/pytorch/aten/src/ATen/native/cuda/Indexing.cu:699: indexSelectLargeIndex: block: [379,0,0], thread: [125,0,0] Assertion `srcIndex < srcSelectDimSize` failed.
/pytorch/aten/src/ATen/native/cuda/Indexing.cu:699: indexSelectLargeIndex: block: [379,0,0], thread: [126,0,0] Assertion `srcIndex < srcSelectDimSize` failed.
/pytorch/aten/src/ATen/native/cuda/Indexing.cu:699: indexSelectLargeIndex: block: [379,0,0], thread: [127,0,0] Assertion `srcIndex < srcSelectDimSize` failed.
Traceback (most recent call last):                                                                                                                                                        
  File "/home/gnc/jtk-svs/mlp-singer_neo/train.py", line 39, in <module>
    main(args)
  File "/home/gnc/jtk-svs/mlp-singer_neo/train.py", line 20, in main
    trainer.train()
  File "/home/gnc/jtk-svs/mlp-singer_neo/trainer/core.py", line 63, in train
    train_dict = self._train_epoch()
  File "/home/gnc/jtk-svs/mlp-singer_neo/trainer/core.py", line 31, in _train_epoch
    pred = self.model(pitch, text)
  File "/home/gnc/anaconda3/envs/happysong_mlp/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/gnc/jtk-svs/mlp-singer_neo/model/core.py", line 28, in forward
    x = self.embed(x)
  File "/home/gnc/anaconda3/envs/happysong_mlp/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/gnc/anaconda3/envs/happysong_mlp/lib/python3.9/site-packages/torch/nn/modules/linear.py", line 103, in forward
    return F.linear(input, self.weight, self.bias)
  File "/home/gnc/anaconda3/envs/happysong_mlp/lib/python3.9/site-packages/torch/nn/functional.py", line 1848, in linear
    return torch._C._nn.linear(input, weight, bias)
RuntimeError: CUDA error: CUBLAS_STATUS_ALLOC_FAILED when calling `cublasCreate(handle)`

```

I think the cause of this problem is the hyper-parameter value of model.json that is tailored to the csd data.

However, I was able to infer that the meaning of num_pitch is the number of pitches, but I do not know the meaning of d_pitch, num_phone, and d_phone. 

If possible, can you tell me all the meanings of the above 3 hyper parameters and model part arguments of model.json?

**Comment by jaketae** [comment]
Hello, thanks for opening this issue. The parameters you mention all have to do with embedding layers.

* `d_pitch`: Embedding dimension for pitch embeddings
* `d_phone`: Embedding dimension for phoneme embeddings
* `num_phone`: Number of phonemes

So the phoneme embedding table will be of size (`num_phone * d_phone`), and ditto the pitch embedding table.

If you are adapting this model to your own data, you may want to check how the lyrics texts are phonemized (G2P). If there are more phonemes than CSD due to language differences, for instance, then this will cause an error. You may want to adjust `num_phone` accordingly.

Let me know if you have more questions!

**Comment by jungtaekyung1** [comment]
From what I understand, d_phone and d_pitch are hyperparameters involved in the dimension of the embedding, and changing them affects the 'quality of the model' and has nothing to do with the error I'm experiencing.

In order to solve the problem I have, I will have to accurately count the unique number of characters in my txt file and write them in num_phone.

Is my understanding correct?

**Comment by jaketae** [comment]
> From what I understand, d_phone and d_pitch are hyperparameters involved in the dimension of the embedding, and changing them affects the 'quality of the model' and has nothing to do with the error I'm experiencing.

This seems correct, though I can't say for sure that it has nothing to do with the error you've encountered. CUDA messages from PyTorch tend to be very opaque, so I recommend that you try running the model with CPU to get a more informative error message.

> In order to solve the problem I have, I will have to accurately count the unique number of characters in my txt file and write them in num_phone.

It depends on how you tokenize and phonemize your input. If you decide to map one character to one token, then yes, that is the strategy you would use. IIRC in my case, I used G2P and mapped one phoneme to one token.

**Comment by jungtaekyung1** [comment]
Then can you tell me how the num_phone:47 of model.json is decided?

**Comment by jaketae** [comment]
If I recall correctly, it comes from the way G2P code is written. You can find the list of phonemes in [`constants.py`](https://github.com/neosapience/mlp-singer/blob/master/data/constants.py). The code path is roughly:

1. [`serialize.py`](https://github.com/neosapience/mlp-singer/blob/master/data/serialize.py): Script to serialize data
2. [`preprocess.py`](https://github.com/neosapience/mlp-singer/blob/master/data/preprocess.py): Module that defines the `Preprocessor` class used for serialization
3. [`g2p.py`](https://github.com/neosapience/mlp-singer/blob/master/data/g2p.py): Module that defines the `encode()` function, which is called by the `Preprocessor`
4. [`constants.py`](https://github.com/neosapience/mlp-singer/blob/master/data/constants.py): Defines the set of onset, coda, and nucleus tokens

If I'm not mistaken, if you add up all the onset, coda, ane nucleus tokens defined in `constants.py`, it totals to 47.

If your custom dataset is also Korean, you can probably use the script without modifying it.

**Comment by jungtaekyung1** [comment]
Thank you for your quick and friendly reply!

I solved the problem I had.

To explain the process of solving the problem, as a result of checking the exact error message by changing the device from cuda to cpu in config/model.json, 

```
IndexError: list index out of range
 ```

was output, and the number of pitches I entered in model.json and I found out that IndexError appeared because the number of pitches read by the model during the encoding process was different.

 I thought I counted the pitches correctly, but I didn't. 

For people like me who will use neosapience/mlp-singer with their own song data, it is recommended to accurately count the number of pitches in the pre-processing process.
